crates/test-utils/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+///     ]
+/// }
+/// fn test_parses(to_parse: &str, expected: i32) -> Result<(), Box<dyn Error>> {
+///     assert_eq!(expected, to_parse.parse()?);
+///     Ok(())
+/// }
+/// ```
+///
+/// Finally, attributes such as as `#[ignore]` may be added to individual tests:
+///
+/// ```
+/// # use std::error::Error;
+/// # use test_utils::param_test;
+/// #
+/// param_test! {
+///     test_parses -> Result<(), Box<dyn Error>>: [
+///         #[ignore] positive: ("21", 21),
+///         negative: ("-17", -17)
+///     ]
+/// }
+/// fn test_parses(to_parse: &str, expected: i32) -> Result<(), Box<dyn Error>> {
+///     assert_eq!(expected, to_parse.parse()?);
+///     Ok(())
+/// }
+/// ```
+#[macro_export]
+macro_rules! param_test {
+    ($func_name:ident -> $return_ty:ty: [
+        $( $(#[$outer:meta])* $case_name:ident: ( $($args:expr),+ )  ),+$(,)?
+    ]) => {
+        mod $func_name {
+            use super::*;
+
+            $(
+                #[test]
+                $(#[$outer])*
+                fn $case_name() -> $return_ty {
+                    $func_name($($args),+)
+                }
+            )*
+        }
+    };
+    ($func_name:ident: [
+        $( $(#[$outer:meta])* $case_name:ident: ( $($args:expr),+ ) ),+$(,)?
+    ]) => {
+        param_test!($func_name -> (): [ $( $(#[$outer])* $case_name: ( $($args),+ ) ),+ ]);
+    };
+}
+
+/// Macro for creating parametrized *asynchronous* tests.
+///
+/// This macro behaves similarly to the [`param_test`] macro, however it must be used with an
+/// `async` function. For convenience, the macro expands the test cases with the `#[tokio::test]`
+/// attribute. If specifying any additional attributes to any test case, it is necessary to
+/// re-specify the `#[tokio::test]` macro for *every* test case.
+///
+/// See [`param_test`] for more information and examples.
+#[macro_export]
+macro_rules! async_param_test {
+    ($func_name:ident -> $return_ty:ty: [
+        $( $(#[$outer:meta])+ $case_name:ident: ( $($args:expr),+ ) ),+$(,)?
+    ]) => {
+        mod $func_name {
+            use super::*;
+
+            $(
+                $(#[$outer])+
+                async fn $case_name() -> $return_ty {
+                    $func_name($($args),+).await
+                }
+            )*
+        }
+    };
+    ($func_name:ident: [
+        $( $(#[$outer:meta])+ $case_name:ident: ( $($args:expr),+ ) ),+$(,)?
+    ]) => {
+        async_param_test!( $func_name -> (): [ $( $(#[$outer])+ $case_name: ($($args),+) ),* ] );
+    };
+
+    ($func_name:ident: [
+        $( $case_name:ident: ( $($args:expr),+ ) ),+$(,)?
+    ]) => {
+        async_param_test!( $func_name -> (): [ $( #[tokio::test] $case_name: ($($args),+) ),* ] );
+    };
+    ($func_name:ident -> $return_ty:ty: [
+        $( $case_name:ident: ( $($args:expr),+ ) ),+$(,)?
+    ]) => {
+        async_param_test!(
+            $func_name -> $return_ty: [ $( #[tokio::test] $case_name: ( $($args),+ ) ),* ]
+        );
+    }
+}
```

Comment:
Suggestion: If I didn't miss some subtleties, the following shorter version should work as well and not suffer from the disadvantage of having to annotate every test with `#[tokio::test]`:
```suggestion
macro_rules! async_param_test {
    ($func_name:ident -> $return_ty:ty: [
        $( $(#[$outer:meta])* $case_name:ident: ( $($args:expr),+ ) ),+$(,)?
    ]) => {
        mod $func_name {
            use super::*;

            $(
                $(#[$outer])*
                #[tokio::test]
                async fn $case_name() -> $return_ty {
                    $func_name($($args),+).await
                }
            )*
        }
    };
    ($func_name:ident: [
        $( $(#[$outer:meta])* $case_name:ident: ( $($args:expr),+ ) ),+$(,)?
    ]) => {
        async_param_test!( $func_name -> (): [ $( $(#[$outer])* $case_name: ($($args),+) ),* ] );
    };
}
```

Reply:
The problem here is that `#[tokio::test]` can be provided options (e.g., `#[tokio::test(start_paused = true)]`), which would cause a duplicate `tokio::test` attribute. We could capture and handle `tokio::test` attributes separately, but although we do use them, their use has been infrequent so far.